### PR TITLE
Agents deployment in linked clusters

### PR DESCRIPTION
**Pull request by filariow**
> Depends on #28 

[Story 1288](https://issues.redhat.com/browse/APPSVC-1288)

On Cluster Environment creation agents are pushed to its Service and Application namespaces.
When a namespace is added to a Cluster Environment Service or Application namespaces list, agents are pushed into it.
When a namespace is deleted from a Cluster Environment Service or Application namespaces list, agents are removed from it.
When a Cluster Environment is deleted, previously pushed agents are removed.

### CODE REVIEW

File: config/agents/svc/rbac/role.yaml [hunk 1]
```diff
@@ -16,6 +16,25 @@ rules:
...
+- apiGroups:
+  - apps
+  resources:
+  - deployments
+  verbs:
+  - get
+  - list
+  - watch
+  - update
+  - patch
```
Comment (dperaza4dustbit) [inline config/agents/svc/rbac/role.yaml:37]: hum... do you need list, watch and patch? are you missing delete?

File: pkg/primaza/constants/doc.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package constants contains Primaza's constants
+package constants
```
Comment (dperaza4dustbit) [inline pkg/primaza/constants/doc.go:18]: No much going on here, why do we need multiple constants files?

File: pkg/primaza/controlplane/doc.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package controlplane contains code for setting up and running Primaza
+// control plane
+package controlplane
```
Comment (dperaza4dustbit) [inline pkg/primaza/controlplane/doc.go:19]: No much going on.

File: pkg/primaza/workercluster/agentapp.go [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+	return nil
+}
+
+func PushApplicationAgent(ctx context.Context, cli *kubernetes.Clientset, namespace string) error {
+	if _, err := createAgentAppDeployment(ctx, cli, namespace); err != nil && !errors.IsAlreadyExists(err) {
+		return err
+	}
+
+	if _, err := createAgentAppService(ctx, cli, namespace); err != nil && !errors.IsAlreadyExists(err) {
```
Comment (dperaza4dustbit) [inline pkg/primaza/workercluster/agentapp.go:48]: why do we need a service for the agents?

File: pkg/primaza/workercluster/agentapp.go [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+        control-plane: primaza-controller-agentapp
+    spec:
+      securityContext:
+        runAsNonRoot: true
+      containers:
+      - command:
+        - /manager
+        args:
+        - --leader-elect
+        image: agentapp:latest
```
Comment (dperaza4dustbit) [inline pkg/primaza/workercluster/agentapp.go:191]: hardcoded? is this temporary? Lets talk about the name of these images in scrum or refinement.

File: pkg/primaza/workercluster/doc.go [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package workercluster contains code for interacting with a worker cluster
+package workercluster
```
Comment (dperaza4dustbit) [inline pkg/primaza/workercluster/doc.go:18]: Not clear to me why we have all these doc.go files

**Review by dperaza4dustbit — COMMENTED** [review]
Interesting approach. I have some questions inline

File: pkg/primaza/constants/doc.go (hunk 2 above)
Comment (filariow) [reply pkg/primaza/constants/doc.go:18]: the `doc.go` is one of the suggested approaches for [documenting packages](https://go.dev/blog/godoc)

File: config/agents/svc/rbac/role.yaml (hunk 1 above)
Comment (filariow) [reply config/agents/svc/rbac/role.yaml:37]: that's the role that will be assigned to the service agent. it just need read access. I will remove update/patch.

File: pkg/primaza/workercluster/agentapp.go (hunk 5 above)
Comment (filariow) [reply pkg/primaza/workercluster/agentapp.go:192]: yes, that's a placeholder for now and will be replaced in next PRs

File: pkg/primaza/workercluster/agentapp.go (hunk 4 above)
Comment (filariow) [reply pkg/primaza/workercluster/agentapp.go:49]: you're right it's not needed, I added this debugging the problems related to leases and forgot it here. nice catch!

File: config/agents/svc/rbac/role.yaml (hunk 1 above)
Comment (dperaza4dustbit) [reply config/agents/svc/rbac/role.yaml:37]: I still see the update and watch there

**Review by dperaza4dustbit — APPROVED** [review]
Since this PR is so big can we merge to get Andy unblocked and then we can do a demo in cabal where we can ask question, my main concern with this PR is giving primaza role to much access to deployments in the target namespaces. We should be able to create a deployment for the agent and maybe delete it, but there is no need to read, list or update any other deployments that could be from the applications

**Comment by filariow** [comment]
@dperaza4dustbit I investigated permissions further and we can scope Get/List/Update/Patch/Delete permission with resource name, so that will be limited to agent deployment. We can't do that for Create, because in kubernetes RBAC happens at URL level. I updated Primaza's permission on namespace resources: https://github.com/primaza/primaza/blob/a8b238f5a0f882f7e3ae6ab60295fcf72566160d/test/acceptance/features/steps/workercluster.py#L165

**Review by dperaza4dustbit — APPROVED** [review]
/lgtm

**Pull request merged**